library/std/src/sys/windows/alloc.rs

```diff
@@ -6,56 +8,81 @@ use crate::sys_common::alloc::{realloc_fallback, MIN_ALIGN};
 struct Header(*mut u8);
 
 unsafe fn get_header<'a>(ptr: *mut u8) -> &'a mut Header {
-    &mut *(ptr as *mut Header).offset(-1)
+    // SAFETY: the safety contract must be upheld by the caller
+    unsafe { &mut *(ptr as *mut Header).offset(-1) }
 }
 
 unsafe fn align_ptr(ptr: *mut u8, align: usize) -> *mut u8 {
-    let aligned = ptr.add(align - (ptr as usize & (align - 1)));
-    *get_header(aligned) = Header(ptr);
-    aligned
+    // SAFETY: the safety contract must be upheld by the caller
```

Comment:
What is the safety contract?

Reply:
I'm not sure how to word this one tbh.

Reply:
Hmm .. maybe something like '`ptr`, once aligned, must have a `header` at `ptr.offset(-1)`'.